pkg/controller/types.go

```diff
@@ -0,0 +1,118 @@
+package controller
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type SynchronizationState uint8
+
+const (
+	SynchronizationStateUnknown SynchronizationState = iota
+	SynchronizationStateSuccessful
+	SynchronizationStateFailed
+)
+
+func (s *SynchronizationState) String() string {
+	switch *s {
+	case SynchronizationStateUnknown:
+		return "Unknown"
+	case SynchronizationStateSuccessful:
+		return "Successful"
+	case SynchronizationStateFailed:
+		return "Failed"
+	}
+	panic(fmt.Sprintf("Invalid SynchronizationState: %d", s))
+}
+
+func (s *SynchronizationState) UnmarshalText(text []byte) error {
+	str := string(text)
+	switch str {
+	case "Successful":
+		*s = SynchronizationStateSuccessful
+	case "Failed":
+		*s = SynchronizationStateFailed
+	case "Unknown":
+		*s = SynchronizationStateUnknown
+	default:
+		return fmt.Errorf("invalid SynchronizationState: %v", str)
+	}
+	return nil
+}
+
+func (s *SynchronizationState) MarshalText() (text []byte, err error) {
+	return []byte(s.String()), nil
+}
+
+// +kubebuilder:object:generate=true
+
+type NaisStatus struct {
```

Comment:
Should we use the status subresource marker to separate the desired state (spec) and observed state (status)?

This is most noticable for resources that are updated by other controllers such as `naiserator` where the status stanza is overwritten (because the existing status is not retrieved), i.e. `PUT` vs `PATCH`.

For sync hash comparisons, this effectively means that every time a resource is updated, a reconciliation is run even if the spec itself has no changes.

Having status as a subresource requires you to be explicit in updating the status (e.g. `r.Client.Status().Update()`), as well as being able to restrict operations to it by means of RBAC separately from the spec stanza.

See also the Kubernetes documentation and design proposal: 
- https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#subresources
- https://github.com/kubernetes/design-proposals-archive/blob/main/api-machinery/customresources-subresources.md

Kubebuilder also recommends it: https://book.kubebuilder.io/reference/generating-crd.html#subresources

Reply:
This makes a lot of sense.
The subresource marker needs to be on the top-level CRD, so wouldn't be here anyway.

On the other hand, while reading these, I realize I've made assumptions that are invalid, so there are some bugs in the reconciler that needs to be fixed ...